### ISSUE TITLE
Group-hook recieved a %fact, but not a %watch-ack

### ISSUE THREAD

**Issue by ixv** [urbit/urbit#2872]
I have a repro, but not a totally from-scratch one. On my ship, when I remove myself from the tlon group and re-add myself, group-hook will send a %watch to ~dopzod, ~dopzod will recieve it and send a %fact back. My ship recieves the %fact, but when I check the bowl, I see the subscription has acked=%.n

cc @philipcmonk

**Comment by belisarius222** [comment]
Could this be an error-handling problem?  I think there's a bug in Gall where it sends a %poke-ack when it should instead kill a subscription.

—
~rovnys-ricfer
https://urbit.org

**Comment by philipcmonk** [comment]
TL;DR if you breach while a gall message is outstanding to another ship, future messages on that wire may receive the wrong sort of acknowledgment.

I spent a while with @ixv's ship and was able to narrow it down a lot.

The useful printf was this in the `%done` case of +mo-handle-sys-way:

```
~?  >  =(wire /way/~dopzod/group-hook)  [hen sys-way-outstanding=(~(get by outstanding.agents.state) [sys+wire hen])]
...
~?  >>  =(wire /way/~dopzod/group-hook)  sys-way-rr=remote-request
```

This gave the following output when leaving the group:

```
>   [~[/g/use/group-hook/~novlud-padtyv/out/~dopzod/group-hook/~dopzod/group/~dopzod/tlon /d //term/1] sys-way-outstanding=[~ [n=%leave l={} r={%watch}]]]
>>  sys-way-rr=p=%watch
%group-hook: got ack for nonexistent subscription
```

And the following when rejoining the group:

```
>   [~[/g/use/group-hook/~novlud-padtyv/out/~dopzod/group-hook/~dopzod/group/~dopzod/tlon /d //term/1] sys-way-outstanding=[~ [n=%leave l={%watch} r={}]]]
>>  sys-way-rr=p=%leave
```

This code is trying to decide what sort of acknowledgment it has received, based on the queue of messages we've sent this ship.  This is a fragile approach since it relies on perfectly lining up the messages we send with the acknowledgments we get back.  If they ever get misaligned, we silently get the wrong sort of acknowledgment.

The astute reader will notice that in this case, leaving the group produces an acknowledgment for a `%watch` and joining it produces an acknowledgment for a `%leave` (which gets swallowed because we don't report `%leave` acknowledgments).  This is exactly backward.  This explains the "got ack for nonexistent subscription" printf and why the subscription remains unacknowledged after rejoining.

It appears that we somehow got an extra item in the "outstanding" queue, and so all further acknowledgments are off by one.

Unfortunately, it's not easy to verify exactly how this happened.  This means somehow we sent a message, received no acknowledgment, but the flow didn't get stuck.  Upon inspection, I note that we do not filter `outstanding` when a ship breaches.  Therefore, my best guess is that ~novlud had sent a message to ~dopzod on this duct/wire and did not hear an acknowledgment before ~dopzod breached.  Since a group with the same name was recreated, the same duct/wire was used, but it already had something in its `outstanding` queue.

The straightforward solution is to filter `outstanding` when a ship breaches.  This is easy to do.

I'm not sure what's the easiest way to fix this problem for existing cases.  A violent approach is to clear the queue entirely once.  There is precedent for this: we did it for `+state-3-to-4` soon after the `outstanding` queue was introduced.  There's already a shim that says if it finds nothing in the queue, it'll simply give both types of acknowledgments.  This is incorrect, but it's unlikely to cause a problem in practice -- a few outstanding pokes and watches will receive *both `watch-ack` and `poke-ack` on their wire, which will likely just crash harmlessly.  Unless I can think of a better option, I'll plan to do this.

Of course, there are also many apparently unacknowledged subscriptions out there.  I think these are basically harmless.  We could kill all these subscriptions expecting them to reconnect, but I don't think this is worth it.

**Comment by vvisigoth** [comment]
As a (I think) sufferer of this, is there any action I can take now or
should I wait for an OTA-d fix?
—
~poldec-tonteg
http://urbit.org

**Comment by philipcmonk** [comment]
Nope, gotta wait for the OTA

**Comment by philipcmonk** [comment]
This should be fixed in the OTA that just went out.  Re-open if this continues to be an issue.
